### PR TITLE
Allow empty collection to insert/insertOrUpdate(Collection)

### LINKED ISSUE #3103: insertOrUpdate(Collection) does not handle empty list properly
#### Goal

> What do you want to achieve?

Save empty list

#### Expected Results

Ignore it considering it's empty

#### Actual Results

     07-01 10:56:02.262 9258-9364/E/BasePersistDbTaskImpl: Error during saving [CompletedRecordsXML].
        java.util.NoSuchElementException
        at java.util.AbstractList$SimpleListIterator.next(AbstractList.java:59)
        at io.realm.internal.modules.CompositeMediator.insertOrUpdate(CompositeMediator.java:121)
        at io.realm.Realm.insertOrUpdate(Realm.java:853)

On this line

        RealmProxyMediator mediator = getMediator(Util.getOriginalModelClass(Util.getOriginalModelClass(objects.iterator().next().getClass())));

#### Steps & Code to Reproduce

Try to save an empty RealmList.

#### Version of Realm and tooling
Realm version(s): 1.1.0

Android Studio version: 2.1.1 but it's irrelevant

Which Android version and device: 6.0.2, Nexus 5X

### PR DESCRIPTION
**Pull request by zaki50**
fixes #3103

- [x] add tests
- [x] fix implementation
- [x] update changelog

### CODE REVIEW

**Comment by zaki50** [comment]
review this please @realm/java

File: realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java [hunk 1]
```diff
@@ -377,6 +384,74 @@ public void insertToRealm_list() {
...
+        try {
+            realm.executeTransaction(new Realm.Transaction() {
+                @Override
+                public void execute(Realm realm) {
+                    realm.insert(Collections.<Cat>emptyList());
+                }
+            });
+            assertEquals(0, realm.where(Cat.class).count());
+        } finally {
+            realm.close();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java]: This is being called in `tearDown()` as well?

File: realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java [hunk 2]
```diff
@@ -106,6 +106,9 @@ public void insert(Realm realm, RealmModel object, Map<RealmModel, Long> cache)
 
     @Override
     public void insert(Realm realm, Collection<? extends RealmModel> objects) {
+        if (objects.isEmpty()) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java]: IMO we should move these checks to the `Realm` class instead as that is where the `null` check already is. I would avoid spreading them out across multiple classes.

File: realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java [hunk 3]
```diff
@@ -118,6 +121,9 @@ public void insertOrUpdate(Realm realm, RealmModel object, Map<RealmModel, Long>
 
     @Override
     public void insertOrUpdate(Realm realm, Collection<? extends RealmModel> objects) {
+        if (objects.isEmpty()) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java]: same

File: realm/realm-library/src/main/java/io/realm/internal/modules/FilterableMediator.java [hunk 4]
```diff
@@ -121,6 +121,9 @@ public void insert(Realm realm, RealmModel object, Map<RealmModel, Long> cache)
 
     @Override
     public void insert(Realm realm, Collection<? extends RealmModel> objects) {
+        if (objects.isEmpty()) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/modules/FilterableMediator.java]: same

File: realm/realm-library/src/main/java/io/realm/internal/modules/FilterableMediator.java [hunk 5]
```diff
@@ -133,6 +136,9 @@ public void insertOrUpdate(Realm realm, RealmModel object, Map<RealmModel, Long>
 
     @Override
     public void insertOrUpdate(Realm realm, Collection<? extends RealmModel> objects) {
+        if (objects.isEmpty()) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/modules/FilterableMediator.java]: same

File: realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java (hunk 1 above)
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java]: No, this Realm is opened in line 410.

File: realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java (hunk 2 above)
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java]: Hm, empty collection is a valid input. Shouldn't every *Mediator be able to handle all valid input correctly?

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java]: The mediator is only for internal use, so we could can decide where the invariant is checked :) ... I don't know if it is possible to move the null check to the mediator, but that would also be an option IMO

Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/modules/CompositeMediator.java]: fixed

**Comment by zaki50** [comment]
review this please @realm/java

**Comment by cmelchior** [comment]
👍

**Comment by zaki50** [comment]
retest this please

File: realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java [hunk 6]
```diff
@@ -134,7 +134,7 @@
      * After being inserted any changes to the original objects will not be persisted.
      *
      * @param realm reference to the {@link Realm} where the objects will be inserted.
-     * @param objects Collection of {@link RealmObject} to insert or update.
+     * @param objects Collection of {@link RealmObject} to insert or update. This must not be empty.
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java]: I think that "must not" is a bit too strong. If `objects` is empty, it is a no-op (we don't throw an exception). Maybe "should not" is better.

**Comment by kneth** [comment]
:+1: with a minor comment on documentation.

File: realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java (hunk 6 above)
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java]: no, if the `objects` is empty, some mediators actually throws an exception.

see https://github.com/realm/realm-java/pull/3106#discussion-diff-69291749

**Comment by zaki50** [comment]
retest this please

**Comment by beeender** [comment]
:+1:

**Comment by zaki50** [comment]
retest this please

**Pull request merged**